text/0000-guaranteed-tco.md

```diff
@@ -0,0 +1,607 @@
+        // Note that this would be a `return accum;`
+        accum
+    }
+}
+```
+
+
+### Use Case 2: Interpreter
+In an interpreter the usual loop is to get an instruction, match on that instruction to find the corresponding function, **call** that function, and finally return to the loop to get the next instruction. (This is a simplified example.)
+
+```rust
+fn exec_instruction(mut self) {
+    loop {
+        let next_instruction = self.read_instr(); // this call can be inlined
+        match next_instruction {
+            Instruction::Foo => self.execute_instruction_foo(),
+            Instruction::Bar => self.execute_instruction_bar(),
+        }
+    }
+}
+```
+
+This example can be turned into the following code, which no longer does any calls and instead just uses jump instructions. (Note that this example might not be the optimal way to use `become`.)
+
+```rust
+fn execute_instruction_foo(mut self) {
+    // foo things ...
+
+    become self.next_instruction();
+}
+
+fn execute_instruction_bar(mut self) {
+    // bar things ...
+
+    become self.next_instruction();
+}
+
+fn next_instruction(mut self) {
+    let next_instruction = self.read_instr(); // this call can be inlined
+    match next_instruction {
+        Instruction::Foo => become self.execute_instruction_foo(),
+        Instruction::Bar => become self.execute_instruction_bar(),
+    }
+}
+```
+
+### Omission of the `become` keyword causes the call to be `return` instead.
+([original example](https://github.com/rust-lang/rfcs/pull/1888#issuecomment-278988088))
+
+This is a potential source of confusion, indeed in a function language where every call is expected to be TCO this would be quite unexpected. (Maybe in functions that use `become` a lint should be applied that enforces usage of either `return` or `become` in functions where at least one `become` is used.)
+
+```rust
+fn foo(x: i32) -> i32 {
+    if x % 2 {
+        let x = x / 2;
+        // one branch uses `become`
+        become foo(new_x);
+    } else {
+        let x = x + 3;
+        // the other does not
+        foo(x) // == return foo(x);
+    }
+}
+```
+
+### Alternating `become` and `return` calls
+([original example](https://github.com/rust-lang/rfcs/pull/1888#issuecomment-279062656))
+
+Here one function uses `become` the other `return`, this is another potential source of confusion. This mutual recursion
+would eventual overflow the stack. As mutual recursion can also happen across more functions, `become` needs to be used
+consistently in all functions if TCO should be guaranteed. (Maybe it is also possible to create a lint for these
+use cases as well.)
+
+```rust
+fn foo(n: i32) {
+    // oops, we forgot become ..
+    return bar(n); // or alternatively: `bar(n)`
+}
+
+fn bar(n: i32) {
+    become foo(n);
+}
+```
+
+
+## Explaining how Rust programmers should *think* about the feature, and how it should impact the way they use Rust. It should explain the impact as concretely as possible.
```

Comment:
These are supposed to be prompts to help fill out the section, not actual headers in the document. Take a look at how some other rfcs at https://rust-lang.github.io/rfcs/ are structured for examples.

Reply:
Can you be more specific? Do you not like the length of the headers, the structure, or something else?

Reply:
They aren't supposed to be headers at all. The whole section should not have any sub-headers (there might be some sub-headers for aspects of the feature but I think they aren't warranted here), and should just be structured as free flow prose which addresses each of these points without literally quoting them in the style of a Q&A.

Reply:
Alright, I'll see what I can do.

Reply:
updated as part of the rewrite to use TCE instead of TCO: https://github.com/rust-lang/rfcs/pull/3407/commits/bef3933b609ea3cef87b0720aec334121ec28fde